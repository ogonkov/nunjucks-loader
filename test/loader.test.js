/* global __USE_ES__ */

import path from 'path';

import compiler from './compiler';
import {statsCompiler} from './stats-compiler';

const loaderBaseOptions = {
    esModule: __USE_ES__
};

describe('Simple compilation', function() {
    test('should compile single files', async function() {
        const output = await compiler('fixtures/single.njk', loaderBaseOptions);

        expect(output({
            title: 'App',
            href: 'https://example.com'
        })).toMatchSnapshot();
    }, /* TODO Find out why it is so slow for first run */ 20000);

    test('should compile templates with inheritance', async function() {
        const output = await compiler('fixtures/child.njk', loaderBaseOptions);

        expect(output({
            parent_link: 'https://example.com/1',
            child_link: 'https://example.com/2'
        })).toMatchSnapshot();
    });

    test('should inherit from parent with super', async function() {
        const output = await compiler('fixtures/inheritance.njk', loaderBaseOptions);

        expect(output({
            name: 'Joe'
        })).toMatchSnapshot();
    });

    test('should compile templates with multiple includes', async function() {
        const output = await compiler('fixtures/multiple-same-includes.njk', loaderBaseOptions);

        expect(output({
            parent_link: 'https://example.com/parent',
            text: 'Button'
        })).toMatchSnapshot()
    });

    test('should compile templates with filters', async function() {
        const output = await compiler('fixtures/builtin-filters.njk', loaderBaseOptions);

        expect(output({
            text: 'nunjucks foolter'
        })).toMatchSnapshot();
    });

    test('should compile builtin tags', async function() {
        const output = await compiler('fixtures/builtin-tags.njk', loaderBaseOptions);

        expect(output({
            tired: true
        })).toMatchSnapshot();
    });

    test('should compile macro', async function() {
        const output = await compiler('fixtures/macro.njk', loaderBaseOptions);

        expect(output()).toMatchSnapshot();
    });

    test('should install Jinja compat', async function() {
        const output = await compiler('fixtures/jinja-syntax.njk', {
            ...loaderBaseOptions,
            jinjaCompat: true
        });

        expect(output({})).toMatchSnapshot();
    })
});

describe('Advanced compilation', function() {
    test('should compile templates with non-relative paths', async function() {
        const output = await compiler('fixtures/django_project/app_example/templates/main/main.njk', {
            ...loaderBaseOptions,
            searchPaths: [
                'test/fixtures/django_project/app_example/templates',
                '.'
            ]
        });

        expect(output()).toMatchSnapshot();
    });

    test('should output warning for dynamic templates imports', async function() {
        const {stats} = await statsCompiler('fixtures/templates/dynamic.njk', {
            ...loaderBaseOptions,
            assetsPaths: [
                'test/fixtures/django_project/app_example/static'
            ]
        });

        expect(stats.toJson().warnings).toContainEqual(expect.objectContaining({
            message: expect.stringContaining('Skipping "test/fixtures/assets/" + name.')
        }));
        expect(stats.toJson().warnings).toContainEqual(expect.objectContaining({
            message: expect.stringContaining('Skipping "test/fixtures/" + macroName + "-macro.njk".')
        }));
    });

    describe('globals', function() {
        const loaderOptions = {
            ...loaderBaseOptions,
            globals: {
                foobar: path.join(__dirname, './fixtures/globals/globals.js')
            }
        };

        test('should compile with given globals', async function() {
            const output = await compiler('fixtures/globals/base.njk', loaderOptions);

            expect(output()).toMatchSnapshot();
        });

        test('should compile globals in parent templates', async function() {
            const output = await compiler('fixtures/globals/child.njk', loaderOptions);

            expect(output()).toMatchSnapshot();
        });

        test('should compile multiple instances of same global', async function() {
            const output = await compiler('fixtures/globals/multiple.njk', loaderOptions);

            expect(output()).toMatchSnapshot();
        })
    });

    describe('extensions', function() {
        const loaderOptions = {
            ...loaderBaseOptions,
            extensions: {
                RemoteExtension: path.join(__dirname, './fixtures/extensions/RemoteExtension.js')
            }
        };

        test('should compile custom tags', async function() {
            const output = await compiler('fixtures/extensions/base.njk', loaderOptions);

            expect(output()).toMatchSnapshot();
        });

        test('should compile custom tags from parent template', async function() {
            const output = await compiler('fixtures/extensions/child.njk', loaderOptions);

            expect(output()).toMatchSnapshot();
        });

        test('should compile multiple instances of same tag', async function() {
            const output = await compiler('fixtures/extensions/multiple.njk', loaderOptions);

            expect(output()).toMatchSnapshot();
        });

        test('should compile async extensions', async function() {
            const output = await compiler('fixtures/extensions/multiple.njk', {
                ...loaderBaseOptions,
                extensions: {
                    RemoteExtension: path.join(__dirname, './fixtures/extensions/RemoteAsyncExtension.js')
                }
            });

            jest.useFakeTimers();

            const asyncRender = output();

            jest.runAllTimers();

            await expect(asyncRender).resolves.toMatchSnapshot();

            jest.useRealTimers();
        });
    });

    describe('filters', function() {
        const loaderOptions = {
            ...loaderBaseOptions,
            filters: {
                foo: path.join(__dirname, './fixtures/filters/foo-filter.js')
            }
        };

        test('should compile single filter instance', async function() {
            const output = await compiler('fixtures/filters/single.njk', loaderOptions);

            expect(output()).toMatchSnapshot();
        });

        test('should compile multiple instances of same filter', async function() {
            const output = await compiler('fixtures/filters/multiple.njk', loaderOptions);

            expect(output()).toMatchSnapshot();
        });

        test('should compile filters in inherited templates', async function() {
            const output = await compiler('fixtures/filters/children.njk', loaderOptions);

            expect(output({
                title: 'Foobar',
                foo_var: 42
            })).toMatchSnapshot()
        });

        test('should compile async filters', async function() {
            const output = await compiler('fixtures/filters/children.njk', {
                ...loaderBaseOptions,
                filters: {
                    foo: path.join(__dirname, './fixtures/filters/foo-filter-async.js')
                }
            });

            jest.useFakeTimers();

            const asyncRender = output({
                title: 'Foobar',
                foo_var: 100500
            });

            jest.runAllTimers();

            const result = await asyncRender;

            expect(result).toMatchSnapshot();

            jest.useRealTimers();
        });
    });
});
