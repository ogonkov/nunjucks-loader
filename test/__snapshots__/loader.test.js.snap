// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Advanced compilation extensions should compile async extensions 1`] = `
"<div id=\\"el100500\\">(async): 
  This content will be replaced with the content from /stuff
</div>


    <div>
        <div id=\\"el100500\\">(async): 
          Content 1
        </div>
    </div>

    <div>
        <div id=\\"el100500\\">(async): 
          Content 2
        </div>
    </div>

"
`;

exports[`Advanced compilation extensions should compile custom tags 1`] = `
"<div id=\\"el100500\\">
  This content will be replaced with the content from /stuff
</div>


"
`;

exports[`Advanced compilation extensions should compile custom tags from parent template 1`] = `
"<div id=\\"el100500\\">
  This content will be replaced with the content from /stuff
</div>


<p>Child template for custom tag check</p>

"
`;

exports[`Advanced compilation extensions should compile multiple instances of same tag 1`] = `
"<div id=\\"el100500\\">
  This content will be replaced with the content from /stuff
</div>


    <div>
        <div id=\\"el100500\\">
          Content 1
        </div>
    </div>

    <div>
        <div id=\\"el100500\\">
          Content 2
        </div>
    </div>

"
`;

exports[`Advanced compilation filters should compile async filters 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"utf-8\\">
    <title>Foobar 100510</title>
  </head>
  <body>
    
    <p>100506</p>

  </body>
</html>
"
`;

exports[`Advanced compilation filters should compile filters in inherited templates 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"utf-8\\">
    <title>Foobar 52</title>
  </head>
  <body>
    
    <p>48</p>

  </body>
</html>
"
`;

exports[`Advanced compilation filters should compile multiple instances of same filter 1`] = `
"15
8
"
`;

exports[`Advanced compilation filters should compile single filter instance 1`] = `
"15
"
`;

exports[`Advanced compilation globals should compile globals in parent templates 1`] = `
"<div>

globals returned value
</div>

Child template

"
`;

exports[`Advanced compilation globals should compile multiple instances of same global 1`] = `
"<div>

globals returned value
</div>

<div>
    Occurrence #1

    
    globals returned value
</div>

<div>
    Occurrence #2

    
    globals returned value
</div>

"
`;

exports[`Advanced compilation globals should compile with given globals 1`] = `
"<div>

globals returned value
</div>

"
`;

exports[`Advanced compilation should compile templates with non-relative paths 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
    <meta charset=\\"UTF-8\\">
    <title></title>
</head>
<body>

    

<main>
    

<p>Main content</p>


</main>



</body>
</html>
"
`;

exports[`Simple compilation should compile builtin tags 1`] = `
"<main>
    <h1><a href=\\"\\">Parent</a></h1>

    
    
      I am tired
    

</main>
"
`;

exports[`Simple compilation should compile macro 1`] = `
"<main>
    <h1><a href=\\"\\">Parent</a></h1>

    
    
<div class=\\"field\\">
  <input type=\\"text\\" name=\\"user\\"
         value=\\"\\" />
</div>

    
<div class=\\"field\\">
  <input type=\\"password\\" name=\\"pass\\"
         value=\\"\\" />
</div>


</main>
"
`;

exports[`Simple compilation should compile single files 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
    <meta charset=\\"UTF-8\\">
    <title>App</title>
</head>
<body>

    <a href=\\"https://example.com\\">Link</a>

</body>
</html>
"
`;

exports[`Simple compilation should compile templates with filters 1`] = `
"<main>
    <h1><a href=\\"\\">Parent</a></h1>

    
    Nunjucks barlter

</main>
"
`;

exports[`Simple compilation should compile templates with inheritance 1`] = `
"<main>
    <h1><a href=\\"https://example.com/1\\">Parent</a></h1>

    
    <p>Child template</p>
    <a href=\\"https://example.com/2\\">Link</a>

</main>
"
`;

exports[`Simple compilation should compile templates with multiple includes 1`] = `
"<main>
    <h1><a href=\\"https://example.com/parent\\">Parent</a></h1>

    
    <p>
        <button>Button</button>

    </p>

    <button>Button</button>


</main>
"
`;

exports[`Simple compilation should inherit from parent with super 1`] = `
"<main>
    <h1><a href=\\"\\">Parent</a></h1>

    
    Super babe

    meets child Joe

</main>
"
`;

exports[`Simple compilation should install Jinja compat 1`] = `
"True
"
`;
